contrib/nexusproto/serializer.go

```diff
@@ -0,0 +1,174 @@
+	// In the second case, we need to instantiate an empty proto struct:
+	if elem.Kind() == reflect.Ptr {
+		empty := reflect.New(elem.Elem())
+		rv.Elem().Set(empty)
+		rv = empty
+	}
+
+	msg, ok := rv.Interface().(proto.Message)
+	if !ok {
+		return nil, fmt.Errorf("%w: value is not a proto.Message or a pointer to one", nexus.ErrSerializerIncompatible)
+	}
+
+	return msg, nil
+}
+
+type protoJsonSerializer struct{}
+
+func (protoJsonSerializer) extractMessageType(contentType string) string {
+	if contentType == "" {
+		return ""
+	}
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil || mediaType != "application/json" || len(params) != 2 || params["format"] != "protobuf" {
+		return ""
+	}
+	return params["message-type"]
+}
+
+func (s protoJsonSerializer) Deserialize(c *nexus.Content, v any) error {
+	messageTypeFromHeader := s.extractMessageType(c.Header["type"])
+	if messageTypeFromHeader == "" {
+		return fmt.Errorf("%w: incompatible content type header", nexus.ErrSerializerIncompatible)
+	}
+	msg, err := messageFromAny(v)
+	if err != nil {
+		return err
+	}
+	messageTypeFromValue := string(msg.ProtoReflect().Type().Descriptor().FullName())
+	if messageTypeFromHeader != messageTypeFromValue {
+		return fmt.Errorf("serialized message type: %q is different from the provided value: %q", messageTypeFromHeader, messageTypeFromValue)
+
+	}
+	return protojson.Unmarshal(c.Data, msg)
+}
+
+func (protoJsonSerializer) Serialize(v any) (*nexus.Content, error) {
+	msg, ok := v.(proto.Message)
+	if !ok {
+		return nil, fmt.Errorf("%w: value is not a proto.Message", nexus.ErrSerializerIncompatible)
+	}
+	data, err := protojson.Marshal(msg)
+	if err != nil {
+		return nil, err
+	}
+	messageType := string(msg.ProtoReflect().Descriptor().FullName())
+	return &nexus.Content{
+		Header: nexus.Header{
+			"type": mime.FormatMediaType("application/json", map[string]string{
+				"format":       "protobuf",
+				"message-type": messageType,
+			}),
+		},
+		Data: data,
+	}, nil
+}
+
+type protoBinarySerializer struct{}
+
+func (protoBinarySerializer) extractMessageType(contentType string) string {
+	if contentType == "" {
+		return ""
+	}
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil || mediaType != "application/x-protobuf" || len(params) != 1 {
+		return ""
+	}
+	return params["message-type"]
+}
+
+func (s protoBinarySerializer) Deserialize(c *nexus.Content, v any) error {
+	messageTypeFromHeader := s.extractMessageType(c.Header["type"])
+	if messageTypeFromHeader == "" {
+		return fmt.Errorf("%w: incompatible content type header", nexus.ErrSerializerIncompatible)
+	}
+	msg, err := messageFromAny(v)
+	if err != nil {
+		return err
+	}
+	messageTypeFromValue := string(msg.ProtoReflect().Type().Descriptor().FullName())
+	if messageTypeFromHeader != messageTypeFromValue {
+		return fmt.Errorf("serialized message type: %q is different from the provided value: %q", messageTypeFromHeader, messageTypeFromValue)
```

Comment:
This is different from how the Go SDK handles proto deserialization, the Go SDK doesn't require the type match

Reply:
I think that may be since the Go SDK added message type later on and the message type header isn't considered required.